serverless/serverless
# Rebuild Credential Handling

## What did you implement:

***Implementing*** [A Potential Fix for] ***Issue***: #1736 



## How did you implement it:

(scoped to AWS)

I migrated a lot of the pre-1.x code forward into 1.x, removing legacy cruft.

Previously you had a number of options, including legacy options for loading credentials.  Given the 0.x=>1.x change, we can drop a lot of the old approaches.  This PR attempts to bring all the good things.

The options for loading credentials are as follows:
1. define credentials on serverless.yml=>service.provider.credentials = { accessKeyId: 'accessKeyId', secretAccessKey: 'secretAccessKey', sessionToken: 'sessionToken' }
2. define a profile from which to get credentials on serverless.yml=>service.provider.profile = 'profile-name' (all profiles loaded using AWS.SharedIniFileCredentials, see http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SharedIniFileCredentials.html)
3. define credentials for all stages using the standard AWS environment variables (see http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/EnvironmentCredentials.html)
4. define a profile for all stages using the environment variable AWS_PROFILE
5. define credentials for each stage using the standard AWS environment variables with the STAGE name inserted (e.g. stage='test', envVarName='AWS_TEST_*')
6. define a profile for each stage using an environment variable `AWS_${stageName.toUpperCase()}_PROFILE`

If credentials/profiles are declared in multiple ways, the later cases will override the former.

These use cases previously covered all user requirements but the current implemenation allows for an expansion of mechanisms if more mechanisms are desirable.



## How can we verify it:



See the tests that were changed and/or created.  That said, the credential over-ridding logic (see docs) is not explicitly tested.

## Todos:



***Is this ready for review?:*** Yes
